### PR TITLE
feat: install command

### PR DESCRIPTION
**Pull request by brendt**
The goal of this PR is to lay the groundwork for installers: classes that can publish files into a user's project. I've already converted the existing `install` command to use the new approach. Another example are the auth migrations.

Let's say this PR gets merged, than we could make an auth installer like so:

```php
final readonly class AuthInstaller implements Installer
{
    use PublishesFiles;

    public function getName(): string
    {
        return 'auth';
    }

    public function publishFiles(): void
    {
        $this->publish(
            source: __DIR__ . 'User.php',
            destination: root_path('src/User.php'),
        );
        
        $this->publish(
            source: __DIR__ . 'UserPermission.php',
            destination: root_path('src/UserPermission.php'),
        );
        
        $this->publish(
            source: __DIR__ . 'CreateUsersTable.php',
            destination: root_path('src/CreateUsersTable.php'),
        );
        
        $this->publish(
            source: __DIR__ . 'CreateUserPermissionsTable.php',
            destination: root_path('src/CreateUserPermissionsTable.php'),
        );
        
        // …
    }
}
```

And run it like so:

```
./tempest install auth
```

Here's a demo of the framework installer:

https://github.com/user-attachments/assets/2d446dd8-4bef-4061-9782-4918ecb973fa

A couple of things to note:

- We cannot hardcode `src`, since some project might have another src directory configured. I think @innocenzi already looked into this, it might even already be available somewhere. We probably need some kind of `src_path` helper as well.
- Currently it only allows for publishing files. Maybe we need more flexibility. @gturpin-dev is working on `make` commands which will depend on this PR. Maybe you can provide some feedback @gturpin-dev ? What am I missing that you'll need?

Overall, would like the input of @aidan-casey , @innocenzi and @gturpin-dev 

TODO:

- [x] Add test helpers to make it easy to create testing install directories, and refactor both the FrameworkInstaller and AuthInstaller tests
- [x] Add finish hook to allow an installer to eg. run migrations

### CODE REVIEW

**Comment by innocenzi** [comment]
I like that a lot!

When the kernel boots, it instantiates the `\Tempest\Support\Composer` class, which holds the main namespace of the user. You can use it to replace the hardcoded `src`:
- https://github.com/tempestphp/tempest-framework/blob/main/src/Tempest/Core/src/Composer.php#L15
- https://github.com/tempestphp/tempest-framework/blob/main/src/Tempest/Core/src/ComposerNamespace.php#L11

It should look like that:
```php
root_path($this->composer->mainNamespace->path);
```

Additionally, we'll still be missing a way to publish files that are supposed to be discovered by Tempest, like migrations. We still need `#[DoNotDiscover]`, so we need to be able to remove it before publishing through `ClassManipulator`. 

While we could do it using the callback, I think it's better to provide a simple API for it in `PublishesFiles`. But in my opinion we should do that in another pull request, to limit the scope of this one

**Review by aidan-casey — COMMENTED** [review]
A couple of thoughts...

**Namespaces**
Somehow we need to deal with namespaces in here. I believe this is what @innocenzi was looking at.

**Allow the installer to specify a discovery location?**
I'm on the fence with this one, but technically we could add something like:

```php
$this->addDiscoveryLocation(__DIR__);
```

The question is does that get away from some of the philosophy or solve any problems we still have today?

**Get name**
It's nice to have the ability to specify the name, but is there a reason it _must_ be set?

**Celebrate the small wins**
We could probably do a better job of celebrating by having fireworks or confetti or something.

```php
$this->showFireworks(seconds: 50);
```


```bash
#!/bin/bash

# Enable colored output
RED='\033[0;31m'
GREEN='\033[0;32m'
YELLOW='\033[0;33m'
BLUE='\033[0;34m'
PURPLE='\033[0;35m'
CYAN='\033[0;36m'
WHITE='\033[0;37m'
RESET='\033[0m'

# Array of colors
colors=("$RED" "$GREEN" "$YELLOW" "$BLUE" "$PURPLE" "$CYAN" "$WHITE")

# Get the terminal window size
rows=$(tput lines)
cols=$(tput cols)

# Function to display fireworks
function fireworks() {
    for ((i=0; i<10; i++)); do
        # Pick a random color
        color=${colors[$RANDOM % ${#colors[@]}]}
        
        # Pick a random explosion center position
        center_row=$((RANDOM % rows))
        center_col=$((RANDOM % cols))

        # Draw the firework explosion in expanding circles
        for radius in {1..5}; do
            # Top, bottom, left, right points
            tput cup $((center_row - radius)) $center_col
            echo -ne "${color}*${RESET}"
            tput cup $((center_row + radius)) $center_col
            echo -ne "${color}*${RESET}"
            tput cup $center_row $((center_col - radius))
            echo -ne "${color}*${RESET}"
            tput cup $center_row $((center_col + radius))
            echo -ne "${color}*${RESET}"

            # Diagonal points
            tput cup $((center_row - radius)) $((center_col - radius))
            echo -ne "${color}*${RESET}"
            tput cup $((center_row - radius)) $((center_col + radius))
            echo -ne "${color}*${RESET}"
            tput cup $((center_row + radius)) $((center_col - radius))
            echo -ne "${color}*${RESET}"
            tput cup $((center_row + radius)) $((center_col + radius))
            echo -ne "${color}*${RESET}"
            
            # Pause between each expansion to simulate explosion
            sleep 0.1
        done
        
        # Clear the firework trail after explosion
        sleep 0.3
        for radius in {1..5}; do
            tput cup $((center_row - radius)) $center_col
            echo -ne " "
            tput cup $((center_row + radius)) $center_col
            echo -ne " "
            tput cup $center_row $((center_col - radius))
            echo -ne " "
            tput cup $center_row $((center_col + radius))
            echo -ne " "
            tput cup $((center_row - radius)) $((center_col - radius))
            echo -ne " "
            tput cup $((center_row - radius)) $((center_col + radius))
            echo -ne " "
            tput cup $((center_row + radius)) $((center_col - radius))
            echo -ne " "
            tput cup $((center_row + radius)) $((center_col + radius))
            echo -ne " "
        done
    done
}

# Clear the screen
clear

# Run the fireworks function
fireworks

# Move cursor to the bottom of the screen
tput cup $rows 0
```

JK 🎆 😛

**Comment by brendt** [comment]
> Additionally, we'll still be missing a way to publish files that are supposed to be discovered by Tempest, like migrations. We still need #[DoNotDiscover], so we need to be able to remove it before publishing through ClassManipulator.

Yes, I remembered this when I was going to bed yesterday as well. We need a way to make changes (remove an attribute, change the namespace). I think it makes sense to have support for that within this PR. Gonna tinker with it

**Comment by brendt** [comment]
> Allow the installer to specify a discovery location?

@aidan-casey I don't see a reason for it. It's already possible by altering the Kernel. We would need a `Kernel::BOOTING` event, to be able to add discovery locations BEFORE the kernel is booted. But that's outside the scope of this PR.

File: src/Tempest/Core/src/PublishesFiles.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        } else {
+            if (! $this->confirm(
+                question: "Do you want to create {$destination}?",
+                default: true,
+            )) {
+                $this->writeln('Skipped');
+
+                return;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (brendt) [reply src/Tempest/Core/src/PublishesFiles.php:40]: There's `--force` built in

File: tests/Integration/Console/Commands/InstallCommandTest.php [hunk 2]
```diff
@@ -22,6 +22,7 @@ protected function setUp(): void
 
         @mkdir($this->installDir());
         chdir($this->installDir());
+        $this->kernel->root = $this->installDir();
```
[the remark this answers is not in the document]
Comment (brendt) [reply tests/Integration/Console/Commands/InstallCommandTest.php:25]: Let's agree that once the filesystem PR is done, we refactor this :)

File: src/Tempest/Core/src/Kernel.php [hunk 3]
```diff
@@ -24,7 +24,7 @@ final class Kernel
     ];
 
     public function __construct(
-        public readonly string $root,
+        public string $root,
```
[the remark this answers is not in the document]
Comment (brendt) [reply src/Tempest/Core/src/Kernel.php:27]: Same as the other comment: when FS is done, we can refactor :)

File: src/Tempest/Framework/Installers/FrameworkInstaller.php [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        );
+
+        $this->publish(
+            source: __DIR__ . '/tempest',
+            destination: root_path('tempest'),
+            callback: function (string $source, string $destination): void {
+                if (PHP_OS_FAMILY !== 'Windows') {
+                    /** @phpstan-ignore-next-line */
+                    exec("chmod +x {$destination}");
+                }
```
[the remark this answers is not in the document]
Comment (brendt) [reply src/Tempest/Framework/Installers/FrameworkInstaller.php:44]: I know, but this isn't the case during a composer install. I've tried this out already and can confirm that we need to manually set it.

File: tests/Integration/Console/Commands/InstallCommandTest.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (aidan-casey) [reply tests/Integration/Console/Commands/InstallCommandTest.php]: Filesystem wouldn't help us here. Simply referencing the testing approach.

**Comment by brendt** [comment]
One thing still missing is the ability to automatically run migrations with the newly added files. I made a separate issue for that: #627

**Comment by gturpin-dev** [comment]
I'm late, but I like this addition 👌 
We could refactor the `PublishFiles` to use the Generator package on a later PR

**Pull request merged**